src/wallet/wallet.cpp

```diff
@@ -3998,9 +3973,18 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();
 
-    auto locked_chain = chain.lock();
     LOCK(walletInstance->cs_wallet);
 
+    // Register wallet with validationinterface. It's done before rescan to avoid
+    // missing block connections between end of rescan and validation subscribing.
+    // Because of wallet lock being hold, block connection notifications are going to
+    // be pending on the validation-side until lock release. It's likely to have
+    // block processing duplicata (if rescan block range overlaps with notification one)
+    // but we guarantee at least than wallet state is correct after notifications delivery.
+    // This is temporary until rescan and notifications delivery are unified under same
+    // interface.
+    walletInstance->m_chain_notifications_handler = walletInstance->chain().handleNotifications(walletInstance);
```

Comment:
In commit "[wallet] Remove locked_chain from CWallet, its RPCs and tests" (ef8c6ca60767cac589d98ca57ee33179608ccda8)

Note I wrote a test to detect the race condition this early `handleNotifications` call prevents: 25651aad58b1f6e543f1ad565d821de46268e724 ([branch](https://github.com/ryanofsky/bitcoin/commits/review.16426.5-edit)). Most of the test works on master so I made a separate PR #18727 for it. There's just a few lines in 25651aad58b1f6e543f1ad565d821de46268e724 not in that PR, helping the test work with this PR